File: pset2/readability/readability.c
#include <stdio.h>
#include <cs50.h>
#include <string.h>
#include <math.h>

int main(void)
{
    string s = get_string("Text: ");

    int letters = 0;
    int words = 0;
    int sentences = 0;

    for (int i = 0; s[i] != '\0'; i++)
    {
        // Add to the letters if its bettwen upper case a and lower case z
        if (s[i] >= 'A' && s[i] <= 'z')
        {
            // assumption that if the first letter is an alphabet character the first word has started
            if (i == 0)
            {
                words++;
            }

            letters++;
        }

        // Add to words if we have a space
        if (s[i] == ' ')
        {
            // make sure next letter is also not a space
            if (s[i + 1] != ' ')
            {
                words++;
            }
        }

        // Add to sentence if we have a . ! or a ?
        if (s[i] == '.' || s[i] == '!' || s[i] == '?')
        {
            sentences++;
        }
    }

    // printf("\n\n\n\n%i letters, %i words, %i sentences\n\n", letters, words, sentences);


    // Calculate the Coleman-Liau index
    float index = 0.0588 * (100.0 * letters / words) - 0.296 * (100.0 * sentences / words) - 15.8;

    // printf("%f\n", index);÷

    if (index < 1)
    {
        printf("Before Grade 1\n");
        return 0;
    }

    if (index >= 16)
    {
        printf("Grade 16+\n");
        return 0;
    }

    printf("Grade %.0d\n", (int) round(index));
}

File: pset6/readability/readability.py
from cs50 import get_string
from math import floor


def main():
    # get the text from the user
    text = get_string("Text: ")

    # put the calculations in a list
    data = get_grade(text)

    # compute the index
    grade = 0.0588 * (100.0 * data["letters"] / data["words"]) - 0.296 * (100.0 * data["sentences"] / data["words"]) - 15.8

    # output the grade
    format_grade(grade)


def get_grade(text):
    letters = 0
    words = 0
    sentences = 0

    # iterate based on the number of characters inputted
    for i in range(len(text)):
        # assume that first item is the start of a word
        if i == 0:
            words += 1

        if text[i] >= "A" and text[i] <= "z":
            letters += 1

        # assume that we have a space its probably the end of one word
        if text[i] == " ":
            # if the next letter is also not a space then its probably safe to say we have another word
            if text[i + 1] != " ":
                words += 1

        # assume a sentence has finished if we have a . ! or ?
        ends = [".", "!", "?"]

        if text[i] in ends:
            sentences += 1

    return {
        'letters': letters,
        'words': words,
        'sentences': sentences
    }


def format_grade(grade):
    if grade <= 1:
        print("Before Grade 1")
    elif grade >= 16:
        print("Grade 16+")
    else:
        print(f"Grade {round(grade)}")


main()

